### PR TITLE
Replace algos for floats and doubles between 0 and 1 in Randomizer

### PR DESCRIPTION
**Pull request by bvenners**
This should go in after #1476.

### CODE REVIEW

**Comment by cheeseng** [comment]
@bvenners @jducoeur  Imho the java version is simpler, it is simply: 

(choose a random between min and max value) / max value

e.g. if the the possible value range is 0 -100, then it is: 

(choose a random between 0 - 100) / 100

which will return a value between 0.00 - 1.00

In the case of float, as in the javadoc: 

return next(24) / ((float)(1 << 24));

If I understand it correctly it means: 

(choose a random between 0 - 2^24, inclusive) / (2^24).toFloat
= 
(choose a random between 0 - 16777216) / 16777216.0

File: scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala [hunk 1]
```diff
@@ -138,8 +138,38 @@ class Randomizer(private[scalatest] val seed: Long) { thisRandomizer =>
...
+    */
+
+    // The total number of possible images for Float between 0.0 and 1.0 is:
+    // (2 ** 23) * (126) + 2
+    // 2 to the power of 23 (0x800000) is how many different mantissas can be represented in 23 bits.
+    // 126 (or 0x7e) is how many exponents there are for numbers between 0.0f and 1.0f, exclusive.
+    // 2 is for 0.0f and 1.0f, which we must handle specially, because their mantissas must be all zeros.
+    // scala> 0x800000 * 0x7e + 2
+    // res14: Int = 1056964610
+    val (x, r) = chooseInt(0, 1056964610)
```
Comment (jducoeur) [inline scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala]: Same comment as the other review: it's a bit-centric number, so I'd write this in hex.

File: scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala [hunk 2]
```diff
@@ -138,8 +138,38 @@ class Randomizer(private[scalatest] val seed: Long) { thisRandomizer =>
...
+
+    // Pick two lucky numbers to play the lotto with:
+    if (x == 333)
+      (0.0f, r)
+    else if (x == 222)
+      (1.0f, r)
```
Comment (jducoeur) [inline scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala]: This is inclusive of 1?  That's actually not what I would expect.  Have you checked that the various uses are comfortable with it?

(It's not obviously wrong, just unintuitive to me.  I would recommend tweaking the ScalaDoc to mention that explicitly.)

File: scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala [hunk 3]
```diff
@@ -166,11 +204,57 @@ class Randomizer(private[scalatest] val seed: Long) { thisRandomizer =>
...
+
+    // The total number of possible images for Double between 0.0 and 1.0 is:
+    // (2 ** 52) * (1022) + 2
+    // 2 to the power of 52 (0x10000000000000L) is how many different mantissas can be represented in 52 bits.
+    // 1022 (or 0x3fe) is how many exponents there are for numbers between 0.0 and 1.0, exclusive.
+    // 2 is for 0.0 and 1.0, which we must handle specially, because their mantissas must be all zeros.
+
+    // scala> 0x10000000000000L * 0x3fe + 2
+    // res20: Long = 4602678819172646914
+    val (x, r) = chooseLong(0, 4602678819172646914L)
```
Comment (jducoeur) [inline scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala]: Usual comment about hex.

File: scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala (hunk 1 above)
Comment (bvenners) [reply scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala]: Will make it hex.

File: scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala (hunk 2 above)
Comment (bvenners) [reply scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala]: Oops that is a bug. It should indeed be exclusive of 1.0. I'll fix that.

File: scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala (hunk 3 above)
Comment (bvenners) [reply scalatest/src/main/scala/org/scalatest/prop/Randomizer.scala]: Will hexify.

**Comment by bvenners** [comment]
I'm going to close this one and pull these changes into the branch being used by PR 1476, make the review suggestions there, and then we can merge all of that in via 1476.

**Pull request closed without merging**